aten/src/ATen/native/cuda/TensorModeKernel.cu

```diff
@@ -0,0 +1,370 @@
+#include <ATen/Context.h>
+#include <ATen/Dispatch.h>
+#include <ATen/TensorIterator.h>
+#include <ATen/native/ReduceOpsUtils.h>
+#include <ATen/native/TensorCompare.h>
+#include <thrust/device_ptr.h>
+#include <thrust/device_vector.h>
+#include <thrust/execution_policy.h>
+#include <thrust/extrema.h>
+#include <thrust/inner_product.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+#include <ATen/cuda/detail/IndexUtils.cuh>
+#include <ATen/native/cuda/SortingCommon.cuh>
+#include <ATen/native/cuda/TensorModeKernel.cuh>
+#include <THC/THCThrustAllocator.cuh>
+
+#define MAX_GRID_SIZE 65535
+#define MAX_BLOCK_SIZE 1024
+
+namespace at {
+namespace native {
+
+template <typename scalar_t>
+void calculate_mode(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    THCudaLongStorage* sort_buffer,
+    THLongStorage* position,
```

Comment:
Got it.
So, I guess `position` can be a `std::vector` (since it's only used in the host), right!?

By the way, there are a bunch of functions in [TensorModeKernel.cuh][1] that use `TH` functions. Should I also get rid of all of them (or bring them to ATen)? For example, it uses:

- `reduceBlockWithNThreadLocalReductions` from [THCReduceApplyUtils.cuh][2];
- `inclusivePrefixScan` from [THCScanUtils.cuh][3];
- `bitonicSortKeys` from [THCSortUtils.cuh][4].

[1]: https://github.com/pytorch/pytorch/pull/52043/files#diff-8327a1711c8c3c8da49165a4d43d5d9a9c13ad521ebac7e7dbda2ee6ce36e076
[2]: https://github.com/pytorch/pytorch/blob/1903b32c35a7e7432d62e43a2d3a679f6f01066d/aten/src/THC/THCReduceApplyUtils.cuh#L128
[3]: https://github.com/pytorch/pytorch/blob/1903b32c35a7e7432d62e43a2d3a679f6f01066d/aten/src/THC/THCScanUtils.cuh#L32
[4]: https://github.com/pytorch/pytorch/blob/1903b32c35a7e7432d62e43a2d3a679f6f01066d/aten/src/THC/THCSortUtils.cuh#L104

Reply:
>So, I guess position can be a std::vector (since it's only used in the host), right!?

Yeah, that's fine.

>By the way, there are a bunch of functions in TensorModeKernel.cuh that use TH functions. Should I also get rid of all of them (or bring them to ATen)? 

I'd say they should be moved to ATen. A `git grep` shows these functions are only used by mode.

Reply:
> I'd say they should be moved to ATen. A git grep shows these functions are only used by mode.

I moved those functions to the header file.
But, after doing so, the previous `mode` implementation broke (since I moved the code).
Therefore, I thought that removing that implementation completely was the way to go.
Is that ok?

That said, I haven't removed the old `mode` implementation for CPUs. Should I do that?

Reply:
Yes, you should remove all the old TH/THC code for `mode`.

Reply:
Got it.